jgrapht-core/src/main/java/org/jgrapht/generate/RandomRegularGraphGenerator.java

```diff
@@ -0,0 +1,276 @@
+     * @throws IllegalArgumentException if degree is greater than number of nodes
+     * @throws IllegalArgumentException if the value "n * d" is odd
+     */
+    public RandomRegularGraphGenerator(int n, int d, Random rng) {
+        if (n < 0) {
+            throw new IllegalArgumentException("number of nodes must be non-negative");
+        }
+        if (d < 0) {
+            throw new IllegalArgumentException("degree of nodes must be non-negative");
+        }
+        if (d > n) {
+            throw new IllegalArgumentException("degree of nodes must be smaller than or equal to number of nodes");
+        }
+        if ((n * d) % 2 != 0) {
+            throw new IllegalArgumentException("value 'n * d' must be even");
+        }
+        this.n = n;
+        this.d = d;
+        this.rng = rng;
+    }
+
+    /**
+     * Generate a random regular graph.
+     *
+     * @param target target graph
+     * @param vertexFactory vertex factory
+     * @param resultMap result map
+     * @throws IllegalArgumentException if target is directed
+     * @throws IllegalArgumentException if "n == d" and graph is simple
+     */
+    @Override
+    public void generateGraph(Graph<V, E> target, VertexFactory<V> vertexFactory, Map<String, V> resultMap) {
+
+        // directed/mixed case
+        if (!target.getType().isUndirected()) {
+            throw new IllegalArgumentException("target graph must be undirected");
+        }
+
+        // simple case
+        if (target.getType().isSimple()) {
+            // no nodes or zero degree case
+            if (this.n == 0 || this.d == 0) {
+                EmptyGraphGenerator<V, E> emptyGraphGenerator = new EmptyGraphGenerator<>(this.n);
+                emptyGraphGenerator.generateGraph(target, vertexFactory, resultMap);
+            }
+
+            else if (this.d == this.n) {
+                throw new IllegalArgumentException("target graph must be simple if 'n == d'");
+            }
+
+            // complete case
+            else if (this.d == (this.n - 1)) {
+                CompleteGraphGenerator<V, E> completeGraphGenerator = new CompleteGraphGenerator<>(this.n);
+                completeGraphGenerator.generateGraph(target, vertexFactory, resultMap);
+            }
+
+            // general case
+            else {
+                generateSimpleRegularGraph(target, vertexFactory);
+            }
+        }
+
+        // non-simple case
+        else {
+            generateNonSimpleRegularGraph(target, vertexFactory);
+        }
+    }
+
+
+    // auxiliary method to check if there are remaining suitable edges
+    // used in generateSimpleRegularGraph(Graph<V, E> target, VertexFactory<V> vertexFactory)
+    private boolean suitable(Set<Map.Entry<Integer, Integer>> edges, Map<Integer, Integer> potentialEdges) {
+        if (potentialEdges.isEmpty()) {
+            return true;
+        }
+
+        Object[] keys = potentialEdges.keySet().toArray();
```

Comment:
Call to array with a zero length integer array, in order to get back an Integer array, and avoid the cast later on.

```
Integer [] keys = potentialEdges.keySet().toArray(new Integer[0]);
```